File: release/npm/lib/checksum.js
"use strict";

const crypto = require("node:crypto");

/**
 * Parses a goreleaser `scythe_<version>_checksums.txt` file into a map of
 * `filename -> lowercase hex sha256`.
 *
 * Format: `<hex-digest>  <filename>` (two spaces), one per line.
 *
 * @param {string} contents
 * @returns {Map<string, string>}
 */
function parseChecksums(contents) {
  const map = new Map();
  for (const line of contents.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    const match = trimmed.match(/^([0-9a-fA-F]{64})\s+\*?(\S+)$/);
    if (!match) continue;
    const [, hash, filename] = match;
    map.set(filename, hash.toLowerCase());
  }
  return map;
}

/**
 * Looks up the expected checksum for `assetFilename`. Throws if the row is
 * missing -- a missing row means the release is malformed and verification
 * must not be silently skipped.
 *
 * @param {Map<string, string>} checksums
 * @param {string} assetFilename
 * @param {string} checksumsUrl
 * @returns {string} lowercase hex sha256
 */
function expectedChecksum(checksums, assetFilename, checksumsUrl) {
  const expected = checksums.get(assetFilename);
  if (!expected) {
    throw new Error(
      `scythe-cli: no checksum entry for "${assetFilename}" in ${checksumsUrl}. ` +
        "The release appears malformed; refusing to install without verification.",
    );
  }
  return expected;
}

/**
 * Computes the lowercase hex sha256 digest of a buffer.
 *
 * @param {Buffer} buffer
 * @returns {string}
 */
function sha256Hex(buffer) {
  return crypto.createHash("sha256").update(buffer).digest("hex");
}

/**
 * Verifies `buffer` against the expected checksum, case-insensitively.
 * Throws naming both hashes and the source URL on mismatch.
 *
 * @param {Buffer} buffer
 * @param {string} expectedHex
 * @param {string} assetUrl
 */
function verifyChecksum(buffer, expectedHex, assetUrl) {
  const actual = sha256Hex(buffer);
  if (actual.toLowerCase() !== expectedHex.toLowerCase()) {
    throw new Error(
      `scythe-cli: checksum mismatch for ${assetUrl}\n` +
        `  expected: ${expectedHex.toLowerCase()}\n` +
        `  actual:   ${actual}`,
    );
  }
}

module.exports = { parseChecksums, expectedChecksum, sha256Hex, verifyChecksum };

File: release/pypi/src/scythe_sql/checksum.py
"""Checksum parsing and verification for goreleaser checksum files."""

import hashlib
import re

from scythe_sql.errors import ScytheSqlError

_ROW_RE = re.compile(r"^([0-9a-fA-F]{64})\s+\*?(\S+)$")


def parse_checksums(contents: str) -> dict[str, str]:
    """Parses a `scythe_<version>_checksums.txt` file into filename -> lowercase hex sha256."""
    result: dict[str, str] = {}
    for raw_line in contents.splitlines():
        line = raw_line.strip()
        if not line:
            continue
        match = _ROW_RE.match(line)
        if not match:
            continue
        digest, filename = match.groups()
        result[filename] = digest.lower()
    return result


class MissingChecksumError(ScytheSqlError):
    """Raised when the checksums file has no row for the requested asset."""


def expected_checksum(checksums: dict[str, str], asset_filename: str, checksums_url: str) -> str:
    """Looks up the expected checksum for `asset_filename`.

    Raises :class:`MissingChecksumError` when the row is missing -- a
    missing row means the release is malformed and verification must not be
    silently skipped.
    """
    expected = checksums.get(asset_filename)
    if expected is None:
        raise MissingChecksumError(
            f"scythe-sql: no checksum entry for '{asset_filename}' in {checksums_url}. "
            "The release appears malformed; refusing to install without verification."
        )
    return expected


def sha256_hex(data: bytes) -> str:
    """Computes the lowercase hex sha256 digest of `data`."""
    return hashlib.sha256(data).hexdigest()


class ChecksumMismatchError(ScytheSqlError):
    """Raised when a downloaded asset's checksum does not match the expected value."""


def verify_checksum(data: bytes, expected_hex: str, asset_url: str) -> None:
    """Verifies `data` against `expected_hex`, case-insensitively.

    Raises :class:`ChecksumMismatchError` naming both hashes and the source
    URL on mismatch.
    """
    actual = sha256_hex(data)
    if actual.lower() != expected_hex.lower():
        raise ChecksumMismatchError(
            f"scythe-sql: checksum mismatch for {asset_url}\n  expected: {expected_hex.lower()}\n  actual:   {actual}"
        )
